### PR TITLE
MDEV-21027: Assertion `part_share->auto_inc_initialized || !can_use_for_auto_inc_init()' failed in ha_partition::set_auto_increment_if_higher

### PR DESCRIPTION
**Pull request by Shunpoco**
- [x] *The Jira issue number for this PR is: MDEV-21027*

## Description
`ha_partition::set_auto_increment_if_higher` expects `part_share->auto_inc_initialized` is true or `can_use_for_auto_inc_init()` is false (but as [the comment of this method](https://github.com/MariaDB/server/blob/10.3/sql/ha_partition.cc#L10577-L10581) saids, it returns false only if we use Spider engine with DROP TABLE or ALTER TABLE query).

However, because `part_share->auto_inc_initialized` becomes true only after all partitions are opened (since 6dce6aecebe6ef78a14cb5c5c5daa8a355551e40), all asserts in `set_auto_increment_if_higher` fail before all partitions are opened except Spider engine with DROP TABLE or ALTER TABLE.
    
Therefore, I added a conditional expression before set_auto_increment_higher is called.

## How can this PR be tested?
```
--source include/have_partition.inc

CREATE TABLE t1 (a INT AUTO_INCREMENT PRIMARY KEY) PARTITION BY HASH (a) PARTITIONS 3;
REPLACE INTO t1 PARTITION (p0) VALUES (3),(8);
 
# Cleanup
DROP TABLE t1;
```
As the test file I created, I checked both InnoDB, Aria, and MyIsam.

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [x] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

### CODE REVIEW

File: mysql-test/suite/parts/t/mdev-21027.test
Comment (Shunpoco) [inline mysql-test/suite/parts/t/mdev-21027.test:9]: Ask for the reviewer: This test code came from the issue ticket. But actually, before this commit, this query also fails for the same reason:
```
REPLACE INTO t1 PARTITION (p0) VALUES (3);
```
And after this commit, this query is executed successfully. On the other hand, the current query is expected to fail (so I added an --error flag).
I think that both queries are suitable for checking the bug.

I am wondering which test code is better for us.

Comment (nayuta-yanagisawa) [reply mysql-test/suite/parts/t/mdev-21027.test:9]: If both are OK for testing, I believe that the simpler one, here `REPLACE INTO t1 PARTITION (p0) VALUES (3)`, is better. Thank you for trying to minimize the test case. This is a great practice!

File: mysql-test/suite/parts/t/mdev-21027.test
Comment (nayuta-yanagisawa) [inline mysql-test/suite/parts/t/mdev-21027.test:7]: MariaDB's default storage engine is MyISAM. So, the engine of the table `t1` above is MyISAM not InnoDB.

**Review by nayuta-yanagisawa — CHANGES_REQUESTED** [review]
Hi, @Shunpoco. Thank you very much for your contribution! The fix itself looks good to me but I have some comments on the test.

Instead of creating a new test file, could you add your test case to `mysql-test/suite/parts/inc/partition_auto_increment.inc` as https://github.com/MariaDB/server/commit/edbf124515b36dc2a5ba40ca469d0d70507734a3 did. You don't have to maintain three tests for different engine. You only need one test because the file is included by tests for each engine.

NOTE that https://github.com/MariaDB/server/commit/edbf124515b36dc2a5ba40ca469d0d70507734a3 has one mistake. The CREATE TABLE statement should be `CREATE OR REPLACE TABLE t1 (...) ENGINE=$engine PARTITION BY HASH(a);`. Hard coding the engine name makes it meaningless to have different test files for different engines.

For the current bug, the statement would be like the following:

```
eval CREATE TABLE t1 (a INT AUTO_INCREMENT PRIMARY KEY) ENGINE=$engine PARTITION BY HASH (a) PARTITIONS 3;
```

**Comment by nayuta-yanagisawa** [comment]
Pushed to bb branch: https://github.com/MariaDB/server/commit/99f7ba2b65f204c771aeb0e6a3281dfd055302ee

**Comment by nayuta-yanagisawa** [comment]
@Shunpoco You can check failing tests [here](http://buildbot.askmonty.org/ci/reports/cross_reference#branch=bb-10.3-MDEV-21027&revision=&platform=&fail_name=&fail_variant=&fail_info_full=&typ=&info=&dt=&limit=100&fail_info_short=) (some CIs are not used for branch protection). Some tests failed but nothing seems to be related to your fix.

File: mysql-test/suite/parts/t/mdev-21027.test
Comment (Shunpoco) [reply mysql-test/suite/parts/t/mdev-21027.test:7]: Thanks, I misunderstood.

**Comment by Shunpoco** [comment]
@nayuta-yanagisawa 
I fixed :pray:
I have one question: There are two patterns of test sets in partition_auto_increment.inc, test set with `if (!$skip_update)` and without it. How do we choose them?

And may I make an another PR to fix your note (edbf124515b36dc2a5ba40ca469d0d70507734a3 has a hard-coding engine name)?

**Review by nayuta-yanagisawa — APPROVED** [review]
Looks good to me. Let's wait for @holyfoot 's review.

**Comment by holyfoot** [comment]
I think just skipping the set_auto_increment_if_higher() for the REPLACE is not a good idea.
Why not make it working just like the INSERT does in that sence?
I'd recommend this as a fix:
--- a/sql/ha_partition.h
+++ b/sql/ha_partition.h
@@ -1408,7 +1408,8 @@ class ha_partition :public handler
       unless we already did it.
     */
     if (!part_share->auto_inc_initialized &&
-        ha_thd()->lex->sql_command == SQLCOM_INSERT &&
+        (ha_thd()->lex->sql_command == SQLCOM_INSERT ||
+         ha_thd()->lex->sql_command == SQLCOM_REPLACE)  &&

**Comment by nayuta-yanagisawa** [comment]
@Shunpoco Thank you for your update. Could you put proper line breaks in the body of the commit message? The headline must be the same as the title of the issue, so any length is acceptable, but the body is supposed to be broken according to git convention (see the commits by the maintainers).

**Comment by nayuta-yanagisawa** [comment]
pushed to bb branch: https://github.com/MariaDB/server/commit/a4c8596a6b2085da6959e87aad97bdc2da958c27

**Review by nayuta-yanagisawa — APPROVED** [review]

**Comment by holyfoot** [comment]
ok to push the recent version.

**Review by nayuta-yanagisawa — CHANGES_REQUESTED** [review]
@Shunpoco I noticed that `tokudb_parts.partition_auto_increment_tokudb` failed. `partition_auto_increment.inc` seems to be included by the test. Could you fix the failure?

http://buildbot.askmonty.org/ci/reports/cross_reference#branch=&revision=a4c8596&platform=&fail_name=&fail_variant=&fail_info_full=&typ=&info=&dt=&limit=100&fail_info_short=

**Comment by Shunpoco** [comment]
OK, please let me check

**Comment by nayuta-yanagisawa** [comment]
Pushed to bb a872145bb1a6eea6ba982d8774642ef284709bde

**Review by nayuta-yanagisawa — APPROVED** [review]
OK to push.

**Pull request merged**
